File: js/euler51.js
var primes = new Set();

function euler50(){
    var sieved = new Array(1000000).fill(false);
    for (let i = 2; i < sieved.length; i++){
        if (!sieved[i]){
            primes.add(i);
            for (let j = 2 * i; j < sieved.length; j += i){
                sieved[j] = true;
            }
        }
    }
    for (let prime of primes){
        if (replace(prime) != null) return prime;
    }
    return 0;
}

function replace(n){
    let digits = n.toString().split('').reverse().map(Number);
    // replace 0, 1, or 2 so at least 8 replacements can be made
    for (let d = 0; d < 3; d++){
        let mask = 0;
        for (let i = 0; i < digits.length; i++){
            if (digits[i] == d) mask += 10 ** i;
        }
        // digit not found
        if (mask == 0) continue;
        // prime is smallest in family
        let size = 1;
        // check if replacing d with d + 1, d + 2, ..., 9
        // results in a prime number
        for (let t = 1; t < 10 - d; t++){
            if (primes.has(n + mask * t)) size++;
        }
        
        // return smallest prime part of family of size 8
        if (size == 8) return n;
    }
    return null;
}

console.log(euler50());

File: go/euler51.go
package main
import (
    "fmt"
    "math"
)

func sieve(n int) []int {
    primes := []int{}
    sieved := make([]bool, n)
    for i := 2; i < n; i++ {
        if !sieved[i] {
            primes = append(primes, i)
            for j := 2 * i; j < n; j += i {
                sieved[j] = true
            }
        }
    }
    return primes
}

func replace(n int, primes map[int]bool) int {
    num := n
    digits := []int{}
    for num > 0 {
        digits = append(digits, num % 10)
        num /= 10
    }
    // generate mask based on occurrences of digit d
    // only create masks for 0, 1, 2 so at least 8 numbers can be made
    // and guarantee that n is smallest prime in family of digit replacements
    for d := 0; d < 3; d++ {
        mask := 0
        for i := 0; i < len(digits); i++ {
            if digits[i] == d {
                mask += int(math.Pow(float64(10), float64(i)))
            }
        }
        if mask == 0 {
            continue
        }
        size := 1
        for t := 1; t < 10 - d; t++ {
            if primes[n + mask * t] {
                size++
            }
        }
        if size == 8 {
            return n
        }
    }
    return 0
}

func main() {
    res := 0
    primes := sieve(1000000)
    primesSet := make(map[int]bool)
    for _, p := range primes {
        primesSet[p] = true
        
    }
    
    for _, p := range primes {
        res = replace(p, primesSet)
        if res != 0 {
            break
        }
    }
    
    fmt.Println(res)
}
